### PR TITLE
Add error handling and networkId parsing middlewares

### PR DESCRIPTION
**Pull request by LogvinovLeon**

### CODE REVIEW

File: ts/src/errors.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+export enum GeneralErrorCodes {
+    VALIDATION_ERROR = 100,
+    MALFORMED_JSON = 101,
+    ORDER_SUBMISSION_DISABLED = 102,
+    THROTTLED = 103,
+}
+
+export const generalErrorCodesReasons: { [key in GeneralErrorCodes]: string } = {
```
Comment (fabioberger) [inline ts/src/errors.ts]: Let's try and start using the convention `to` for mappings: `generalErrorCodeToReason`

File: ts/src/errors.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+export class NotImplementedError extends RelayerBaseError {
+    public statusCode = 501;
+}
+
+export enum GeneralErrorCodes {
+    VALIDATION_ERROR = 100,
+    MALFORMED_JSON = 101,
+    ORDER_SUBMISSION_DISABLED = 102,
+    THROTTLED = 103,
```
Comment (fabioberger) [inline ts/src/errors.ts]: Convention: camel-case key values.

File: ts/src/middleware/error_handling.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        if (relayerError.statusCode === HttpStatus.BAD_REQUEST) {
+            const badRequestError = relayerError as BadRequestError;
+            if (badRequestError.generalErrorCode === GeneralErrorCodes.VALIDATION_ERROR) {
+                const validationError = badRequestError as ValidationError;
+                const errorBody = {
+                    code: badRequestError.generalErrorCode,
+                    reason: generalErrorCodesReasons[badRequestError.generalErrorCode],
+                    validationErrors: validationError.validationErrors,
+                };
+                res.status(relayerError.statusCode).send(errorBody);
```
Comment (fabioberger) [inline ts/src/middleware/error_handling.ts:38]: Let's `return` here.

File: ts/src/middleware/url_params_parsing.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ */
+export function urlParamsParsing(req: express.Request, _res: express.Response, next: express.NextFunction): void {
+    const networkId = parseNetworkId(req.query.networkId);
+    req.networkId = networkId;
+    next();
+}
+
+function parseNetworkId(networkIdStrIfExists?: string): number {
+    if (_.isUndefined(networkIdStrIfExists)) {
+        return NETWORK_ID;
```
Comment (fabioberger) [inline ts/src/middleware/url_params_parsing.ts]: Rename to `DEFAULT_NETWORK_ID`

File: ts/src/middleware/url_params_parsing.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    } else {
+        const networkId = _.parseInt(networkIdStrIfExists);
+        if (networkId !== NETWORK_ID) {
+            const validationErroItem = {
```
Comment (fabioberger) [inline ts/src/middleware/url_params_parsing.ts]: Typo: `Erro`

**Review by fabioberger — CHANGES_REQUESTED** [review]

File: package.json [hunk 6]
```diff
@@ -28,6 +28,7 @@
     "dependencies": {
         "0x.js": "^1.0.4",
         "@0xproject/connect": "^2.0.2",
+        "@0xproject/order-watcher": "^2.0.0",
```
Comment (fabioberger) [inline package.json:31]: Latest version of all packages has changed. Order-watcher is now `2.1.0` for example. Not sure it's worth upgrading yet, but maybe at the end.

File: ts/src/middleware/url_params_parsing.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+import { DEFAULT_NETWORK_ID } from '../config';
+import { ValidationError } from '../errors';
+
+/**
+ * Parses URL params and stores them on the request object
+ */
+export function urlParamsParsing(req: express.Request, _res: express.Response, next: express.NextFunction): void {
+    const networkId = parseNetworkId(req.query.networkId);
+    req.networkId = networkId;
```
Comment (fabioberger) [inline ts/src/middleware/url_params_parsing.ts:13]: Add a `// HACK:` comment here and explain that this is the recommended way of passing data, despite it not being type-safe and hacky. Source link would be nice too.

**Review by fabioberger — APPROVED** [review]

File: package.json (hunk 6 above)
Comment (LogvinovLeon) [reply package.json:31]: I'll upgrade them as I need those new features and at the end.

File: ts/src/middleware/url_params_parsing.ts (hunk 7 above)
Comment (LogvinovLeon) [reply ts/src/middleware/url_params_parsing.ts:13]: It's hacky, but type safe in a way. I overrode the type of Request in globals and it now has that property. So after the middleware - the code is type-safe.

**Pull request merged**